palette/src/oklch.rs

```diff
@@ -65,31 +48,17 @@ pub struct Oklch<T = f32> {
     pub hue: OklabHue<T>,
 }
 
-impl<T> Copy for Oklch<T> where T: Copy {}
-
-impl<T> Clone for Oklch<T>
+impl<T> Oklch<T>
 where
-    T: Clone,
+    T: Real,
```

Comment:
Unlike the other trait bounds, these are not accidents. Unlike, say RGB colors, the Ok color spaces make no sense based on integers. I'd even go so far as to claim that many computations would have plain wrong results, if integers were used. So I put these trait bounds here to actively prevent misunderstandings

Reply:
Yeah, this library used to have a lot more of those traits specified up-front and on the structs, but I learned that it causes problems (other than being very verbose) and somewhat recently removed them. Ignoring that it blocks `const` in its current form, it put an unnecessary restriction on methods that shouldn't need it. The `Real` trait is a bit special, since it's sort of a classification and less of a practical capability (unlike `Add`, for example), so I can see that it feels more natural to blanket apply it but it still follows the same logic as more capability oriented traits. It's not necessary to be able to do `T + T` to create a new `Oklch<T>` object, but it's required for converting from `Srgb<T>`.

> I'd even go so far as to claim that many computations would have plain wrong results, if integers were used.

Yes, they would. That's why they should require the `Real` trait if it's essential for them. But it's not essential for everything and those other things can allow the "wrong" type to be used, since it does no harm. It may even be helpful in some special cases that we don't even know about. Maybe someone need to use it as a container for some other type and not perform those specific operations in that stage. It's fine to let some odd cases through if it opens up for more (still correct) options. The fact that we limit where it really matters will still prevent errors, just in a different place.

Sorry for the long response but I hope it explains my reasoning.

Reply:
Your reasoning is actually convincing for a public library (and not even long winded :))

If I'm understanding you right, what you are saying in practice comes down to: "conversions should have the trait bounds but all other operations should not." Right? Does that apply only to conversions to and from sRGB or also to and from XYZ and between the Ok* color spaces?

Reply:
Conversion was just an example. It's any function or trait that would be impossible to implement correctly without it. A very strong hint for `Real` in particular is if it involves non-integer constants! There are also some case, like interpolating, where it may still make sense to add even if it's not always necessarily required to make it compile. Simply because the operation itself is nonsense otherwise. They are still limited to that one trait or function.

Reply:
On the subject of integer constants, e.g. `2.0`. Currently I'd use `T:from_f64(2.0)`. Should I use `T::One()+T::One()` instead?

Reply:
No, I would say prefer `from_f64` for simple constant like that. Calculating it is more for when precision may be important. The places where it's done for values like 2.0 is legacy from when I tried to completely avoid converting from floats. I have given up on that. 😅

Reply:
Or, actually, you can also use `T::One()+T::One()` to avoid adding `Real` as a requirement. But if it's there anyway, you might as well use it!